cargo-dist/src/backend/installer/docker.rs

```diff
@@ -0,0 +1,224 @@
+        // (soft-blocked on doing a sysdeps refactor)
+        let release = dist.release(self.release_idx);
+        release
+            .system_dependencies
+            .apt
+            .clone()
+            .into_iter()
+            .filter(|(_, package)| package.0.wanted_for_target(&self.target))
+            .filter(|(_, package)| package.0.stage_wanted(&DependencyKind::Run))
+            .map(|(name, spec)| {
+                if let Some(version) = spec.0.version {
+                    format!("{name}={version}")
+                } else {
+                    name
+                }
+            })
+            .collect()
+    }
+
+    /// Compute the data for the dockerfile
+    ///
+    /// Split off from rendering in case it's useful for testing.
+    fn compute_dockerfile(
+        &self,
+        dist: &DistGraph,
+        manifest: &DistManifest,
+    ) -> DistResult<DockerfileInfo> {
+        let apt_deps = self.apt_runner_deps(dist, manifest);
+
+        let bins = self
+            .bins
+            .iter()
+            .map(|file_name| Bin {
+                source: format!("./{file_name}"),
+                name: file_name.clone(),
+            })
+            .collect();
+        let runner = RunnerImage {
+            image: RUNNER_IMAGE.to_owned(),
+            apt_deps,
+            includes: self.includes.clone(),
+        };
+        Ok(DockerfileInfo { bins, runner })
+    }
+
+    /// Render the dockerfile with the given data
+    fn render_dockerfile(&self, templates: &Templates, info: &DockerfileInfo) -> DistResult<()> {
+        let contents = templates.render_file_to_clean_string(TEMPLATE_INSTALLER_DOCKER, &info)?;
+        let dockerfile_path = self.package_dir.join(DOCKERFILE_NAME);
+        LocalAsset::write_new(&contents, dockerfile_path)?;
+        Ok(())
+    }
+
+    /// Build and export the docker image to a tarball with appropriate metadata,
+    /// assuming `render_dockerfile` has already run.
+    fn build_and_export_image(&self, dist: &DistGraph) -> DistResult<()> {
+        let Some(docker) = &dist.tools.docker else {
+            unreachable!("docker didn't exist, tasks.rs was supposed to catch that!");
+        };
+        // 1. check if the cargo-dist docker backend is setup
+        // 2. create it if not
+        // 3. do a build with the cargo-dist backend
+        //
+        // We do this because the kind of build+export we want to do requires a
+        // "docker-container" buildx builder, which doesn't exist by default. So
+        // we need to create one, and then use it. Unfortunately this is persistent
+        // and non-idempotent, so we need to check if we've already done it first.
+        //
+        // FIXME: is there any way to do this that doesn't require persistently modifying
+        // the current docker install?
+        let has_dist_builder = {
+            let mut cmd = Command::new(&docker.cmd);
+            cmd.arg("buildx")
```

Comment:
Missed earlier - one or more of these commands is going to write to stdout, which interferes with the build manifest that we *also* write to stdout. We ignore it in cargo builds (because `cargo build` exclusively outputs to stderr), but in generic builds I'm very cautious about avoiding writing to stdout. What I do is this:

https://github.com/axodotdev/cargo-dist/blob/main/cargo-dist/src/generic_build.rs#L102-L103

In other words, I allow stderr to get streamed directly to the parent process's stdout, as normal, but capture stdout to a pipe. At the end of the build, I capture and then print that stdout:

https://github.com/axodotdev/cargo-dist/blob/main/cargo-dist/src/generic_build.rs#L135-L139

https://github.com/axodotdev/cargo-dist/blob/main/cargo-dist/src/generic_build.rs#L159-L160

You'll likely want to do something similar for every command you run here.